cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/SparkPipelineRunner.java

```diff
@@ -378,6 +356,123 @@ public void runPipeline(PipelinePhase pipelinePhase, String sourcePluginType,
     }
   }
 
+  private SparkCollection<Object> handleAutoJoin(AutoJoiner autoJoiner, AutoJoinerContext autoJoinerContext,
+                                                 Map<String, SparkCollection<Object>> inputDataCollections) {
+    JoinDefinition joinDefinition = autoJoiner.define(autoJoinerContext);
+
+    // join required stages first to cut down the data as much as possible
+    List<JoinStage> requiredStages = joinDefinition.getStages().stream()
+      .filter(s -> s.isRequired())
+      .collect(Collectors.toList());
+    List<JoinStage> optionalStages = joinDefinition.getStages().stream()
+      .filter(s -> !s.isRequired())
+      .collect(Collectors.toList());
+    List<JoinStage> orderedStages = new ArrayList<>(requiredStages.size() + optionalStages.size());
+    orderedStages.addAll(requiredStages);
+    orderedStages.addAll(optionalStages);
+
+    Iterator<JoinStage> stageIter = orderedStages.iterator();
+    JoinStage left = stageIter.next();
+    SparkCollection<Object> leftCollection = inputDataCollections.get(left.getStageName());
+    Schema leftSchema = left.getSchema();
+
+    JoinCondition condition = joinDefinition.getCondition();
+    // currently this is the only operation, but check this for future changes
+    if (condition.getOp() != JoinCondition.Op.KEY_EQUALITY) {
+      throw new IllegalStateException("Unsupport join condition operation " + condition.getOp());
+    }
+    JoinCondition.OnKeys onKeys = (JoinCondition.OnKeys) condition;
+    Map<String, List<String>> stageKeys = onKeys.getKeys().stream()
+      .collect(Collectors.toMap(j -> j.getStageName(), j -> j.getFields()));
```

Comment:
Use method reference instead of lambda `toMap(JoinKey::getStageName, JoinKey::getFields)`